### DISCUSSION TITLE
GraalVM + RoutingServlet & StaticServlet

### DISCUSSION THREAD

**Discussion by jacksonsalopek** [activej/activej#302] [Q&A] [answered]
Hi,

I'm trying to create a GraalVM native image and it seems like RoutingServlet and StaticServlet are having problems when attempting to run the native image. The repository in question is here: https://github.com/jacksonsalopek/site

Essentially, it seems like a lot of the reflection metadata has been stripped out and is unable to inject dependencies properly. I could be entirely wrong, but I'm not sure why I a) get no logs when running the native image and b) index route seems to load, but static assets are not served and other routes besides the index route throw an Internal Server Error.

I can provide more logs if needed, but the behavior can be replicated by cloning the repo linked above and running `sbt compile nativeImage` then running the executable that is built.

**Comment by eduard-vasinskyi** [comment, marked as answer]
Hi, @jacksonsalopek

By default, the `native-image` tool does not include any resources in the executable. You have to manually configure it to include those resources. To do this, first, create the `resource-config.json` in the site/config/ directory. Then add the following lines to the file:
```json
{
 "resources": {
   "includes": [
     {
       "pattern": ".*"
     }
   ]
 }
}
```
You can fine-tune the resources you want to include. [Here](https://docs.oracle.com/en/graalvm/enterprise/20/docs/reference-manual/native-image/Resources/#accessing-resources-in-native-images) is the tutorial.
Then you need to specify the path to `resource-config.json` using the `H:ResourceConfigurationFiles` option, just as you did with `H:ReflectionConfigurationFiles`.

Regarding logback, there are some reflection issues with newer versions of `logback`. I believe there are [ways to configure](https://stackoverflow.com/questions/72770461/graalvm-native-image-can-not-compile-logback-dependencies) `native-image` to work properly with logback, but for now I just downgraded logback to version `1.3.4` and removed `logback.xml` file from the resources. Configuring `logback` for `native-image` goes beyond the scope of ActiveJ.

When launching the server, I also got several errors when accessing various endpoints. Those errors, I believe,  were related to the initialization of Scala classes. [Adding the `--initialize-at-build-time` option](https://github.com/oracle/graal/issues/1490) to `nativeImageOptions` seemed to help with these errors. I’m not familiar with Scala, perhaps you can find a better solution.

**Reply by jacksonsalopek** [reply]
Hi @eduard-vasinskyi,

All of those steps worked perfectly for me, native image was built correctly and runs properly! I really appreciate the help!!
